### ISSUE TITLE
Tracker causes warning in log: Cannot modify header information - headers already sent

### ISSUE THREAD

**Issue by jaxtheking** [Smile-SA/elasticsuite#1860]
My log has become riddled with warnings such as:
`main.ERROR: Warning: Cannot modify header information - headers already sent by (output started at /home/wwwharri/public_html/dev/vendor/laminas/laminas-http/src/PhpEnvironment/Response.php:115) in /home/wwwharri/public_html/dev/vendor/magento/framework/Stdlib/Cookie/PhpCookieManager.php on line 148 [] []`

I've taken some time to investigate this, disabling literally all third-party modules, one at a time. I've traced it back to the Hit controller of the Tracker module (Smile\ElasticsuiteTracker\Controller\Tracker\Hit), where a call to a tracking pixel is made:
```
public function execute()
{
    $this->getResponse()->setBody(base64_decode(self::PIXEL));
    $this->getResponse()->setHeader('Content-Type', 'image/png');
    $this->getResponse()->sendResponse();

    $eventData = $this->decodeParams($this->getRequest()->getParams());
     $this->trackingService->addEvent($eventData);
 }
```

If I comment out `sendResponse()`, the warning stops being produced.
I can't be sure but it may have started to happen recently enough, after the upgrade to 2.3.4 or 2.3.5.

I'm def no expert, but I thought this would be an issue with the full-page cache - however I checked and the PageCache module is indeed intercepting the call and calling `sendVary()` before `sendResponse()`.
At this point, my knowledge ends and I'm not sure how to debug this any further. I'm hoping you guys can reproduce this issue and shed some light on it. Let me know if you need any more details.

### Preconditions

PHP 7.3 (or 7.2), Apache 2.4 + MPM Event, built-in cache with or without Redis

Magento Version : CE 2.3.5

ElasticSuite Version : 2.9.1, ElasticSearch 7.8.0

Environment : it occurs in both Developer and Production modes

Third party modules : few - but issue still occurs after having disabled them all

**Comment by androshchuk** [comment]
Hello @jaxtheking 

I can't reproduce it in my environment.
This can happen when you printing something before sending headers.
For more information: https://github.com/magento/magento2/issues/14818
Try to check it out.

Regards

**Comment by jaxtheking** [comment]
Thanks for replying @androshchuk 
I'm pretty sure nothing is getting printed from my template (I disabled it and switched to Luma), my modules (disabled) or third-party modules (disabled all but ElasticSuite).

I just realised that, oddly, it starts to happen once a user logs into the site - then it persists after they log out.
However, I noticed it stops if all cookies are cleared... so I then started removing them one after another one and it seems the cause is the `persistent_shopping_cart` cookie. After l remove it (and the user is logged out), the error gets logged on one more page view then it stops.

Could you do another test by logging in as a user in the frontend to see if you're able to reproduce?
Thanks,
Luca

**Comment by androshchuk** [comment]
Unfortunately, I can't reproduce this error. 
@rbayet could you look at this problem?

Regards

**Comment by TomMeijnaerts** [comment]
i have the same problem, after disabling the tracking error disappears ... but thats not a solution ofcourse

````
[2020-07-28 14:42:53] main.CRITICAL: Warning: Cannot modify header information - headers already sent by (output started at vendor/laminas/laminas-http/src/PhpEnvironment/Response.php:115) in /vendor/magento/framework/Stdlib/Cookie/PhpCookieManager.php on line 148 

{"report_id":"a0f0da8c9b478e9987696554af9fd58416e03d205310cca1c589f8b001cfea86","exception":"[ob
ject] (Exception(code: 0): Warning: Cannot modify header information - headers already sent by (output started at vendor/laminas/laminas-http/src/PhpEnvironment/Response.php:115) i
n vendor/magento/framework/Stdlib/Cookie/PhpCookieManager.php on line 148 at vendor/magento/framework/App/ErrorHandler.php:61
)"} []
````

**Comment by jaxtheking** [comment]
The error seems to have disappeared after upgrading to Magento 2.4.0 and ElasticSuite 2.10.0.
Feel free to close the issue, unless others are still seeing this error in their logs.
Thanks for looking into this anyway!

**Comment by tig-dennisvanderhammen** [comment]
The error seems to be caused by core Magento. This happens when you try to change the headers within a controller. Magento themselves caused an undefined error, causing problems when attempting to change the headers. Elasticsearch does this in the code mentioned by @jaxtheking.

As not everybody is able to update to Magento 2.4 immediately, I've created a patch, based on code changes between Magento 2.3 and Magento 2.4. To apply this patch, you can use the following documentation page. https://devdocs.magento.com/guides/v2.4/comp-mgr/patching/composer.html

Please don't forget to rename the .txt file to a .patch file, as Github doesn't support .patch files.

[arealist_frontname.txt](https://github.com/Smile-SA/elasticsuite/files/5621608/arealist_frontname.txt)

**Comment by raybog** [comment]
We think we are getting the same issue (only in production mode):

`{"0":"Warning: Cannot modify header information - headers already sent by (output started at \/app\/vendor\/laminas\/laminas-http\/src\/PhpEnvironment\/Response.php:132) in \/app\/vendor\/magento\/framework\/Stdlib\/Cookie\/PhpCookieManager.php on line 148","1":"#1 setcookie() called at [vendor\/magento\/framework\/Stdlib\/Cookie\/PhpCookieManager.php:148]\n#2 Magento\\Framework\\Stdlib\\Cookie\\PhpCookieManager->setCookie() called at [vendor\/magento\/framework\/Stdlib\/Cookie\/PhpCookieManager.php:101]\n#3 Magento\\Framework\\Stdlib\\Cookie\\PhpCookieManager->setSensitiveCookie() called at [vendor\/magento\/framework\/App\/Response\/Http.php:114]\n#4 Magento\\Framework\\App\\Response\\Http->sendVary() called at [vendor\/magento\/module-page-cache\/Model\/App\/Response\/HttpPlugin.php:25]\n#5 Magento\\PageCache\\Model\\App\\Response\\HttpPlugin->beforeSendResponse() called at [vendor\/magento\/framework\/Interception\/Interceptor.php:121]\n#6 Magento\\Framework\\App\\Response\\Http\\Interceptor->Magento\\Framework\\Interception\\{closure}() called at [vendor\/magento\/framework\/Interception\/Interceptor.php:153]\n#7 Magento\\Framework\\App\\Response\\Http\\Interceptor->___callPlugins() called at [generated\/code\/Magento\/Framework\/App\/Response\/Http\/Interceptor.php:26]\n#8 Magento\\Framework\\App\\Response\\Http\\Interceptor->sendResponse() called at [vendor\/magento\/framework\/App\/Bootstrap.php:262]\n#9 Magento\\Framework\\App\\Bootstrap->run() called at [pub\/index.php:40]\n"`

"smile/elasticsuite": "2.9.0"
"magento" : "2.3.6"

**Comment by jaxtheking** [comment]
As mentioned [above](https://github.com/Smile-SA/elasticsuite/issues/1860#issuecomment-672460722), the error disappeared for me after upgrading to Magento 2.4.0 and ElasticSuite 2.10.0.
I'm not sure which one did it, but I'd recommend trying upgrading both first.

**Comment by tig-dennisvanderhammen** [comment]
@jaxtheking, @raybog, the update to ElasticSuite 2.10.0 fixed it, but unfortunately you cannot install this on 2.3.*. I've created a [patch](https://github.com/Smile-SA/elasticsuite/issues/1860#issuecomment-736355557) if people do need a fix on Magento 2.3.*:

**Comment by romainruaud** [comment]
Okay I close this one then, since it seems it was a Magento bug, and it's also resolved with 2.4.0

Regards
